### ISSUE TITLE
Server crash when opening timeline tab for big workflows

### ISSUE THREAD

**Issue by amitm02** [argoproj/argo-workflows#6049]
When opening the timeline tab (URL /workflows/argo/<workflow>?tab=timeline) for big workflows (~3k nodes),
Argo server crashes:

```
time="2021-05-31T10:45:33.047Z" level=info msg="finished streaming call with code NotFound" error="rpc error: code = NotFound desc = workflows.argoproj.io \"do-pipline-ztr2w\" not found" grpc.code=NotFound grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-05-31T10:45:33Z" grpc.time_ms=4.47 span.kind=server system=grpc
time="2021-05-31T10:45:36.187Z" level=info msg="finished streaming call with code Canceled" error="rpc error: code = Canceled desc = context canceled" grpc.code=Canceled grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-05-31T10:30:11Z" grpc.time_ms=925019 span.kind=server system=grpc
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5a4482]
time="2021-05-31T10:45:33.047Z" level=info msg="finished streaming call with code NotFound" error="rpc error: code = NotFound desc = workflows.argoproj.io \"do-pipline-ztr2w\" not found" grpc.code=NotFound grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-05-31T10:45:33Z" grpc.time_ms=4.47 span.kind=server system=grpc
time="2021-05-31T10:45:36.187Z" level=info msg="finished streaming call with code Canceled" error="rpc error: code = Canceled desc = context canceled" grpc.code=Canceled grpc.method=WatchWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-05-31T10:30:11Z" grpc.time_ms=925019 span.kind=server system=grpc
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5a4482]

goroutine 29074 [running]:
bufio.(*Writer).Flush(0xc00115fb00, 0x1, 0x7f4e459ec698)
	/usr/local/go/src/bufio/bufio.go:607 +0x62
net/http.(*response).Flush(0xc0005089a0)
	/usr/local/go/src/net/http/server.go:1673 +0x35
github.com/argoproj/pkg/grpc/http.writeKeepalive(0x3442c00, 0xc0005089a0, 0xc002198088)
	/go/pkg/mod/github.com/argoproj/pkg@v0.8.1/grpc/http/forwarders.go:151 +0x173
github.com/argoproj/pkg/grpc/http.keepalive(0x344b840, 0xc000658a50, 0x3442c00, 0xc0005089a0, 0xc002198088)
	/go/pkg/mod/github.com/argoproj/pkg@v0.8.1/grpc/http/forwarders.go:166 +0xa5
created by github.com/argoproj/pkg/grpc/http.withKeepalive
	/go/pkg/mod/github.com/argoproj/pkg@v0.8.1/grpc/http/forwarders.go:174 +0x9a
	
```
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
Please can you include pod manifest.

**Comment by alexmt** [comment]
We've got the same error in Argo CD: https://github.com/argoproj/argo-cd/issues/6741


Fix is available in https://github.com/argoproj/pkg/pull/43 . Workflow need to ` go get github.com/argoproj/pkg` to pull in the fix.

**Comment by zhichenghe** [comment]
according to https://github.com/argoproj/argo-workflows/blob/master/CHANGELOG.md,  this issue already been fixed on version of v3.2.0-rc1 (2021-08-19), but we still meet the same issue on the version of v3.2.7.

**Comment by clementlecorre** [comment]
I have a same issue in `v3.2.7`

**Comment by agilgur5** [comment]
> but we still meet the same issue on the version of v3.2.7.

> I have a same issue in `v3.2.7`

There appears to have been a second crash fixed in the same area of the code in argoproj/pkg#47.

That seems to have gotten pulled in with #8439, which was released with v3.4.0-rc1.

Please file a new issue if this is reproduced in 3.4.0+
